File: protocol/api/src/main/java/ink/reactor/protocol/api/connection/PlayerConnection.java
package ink.reactor.protocol.api.connection;

import ink.reactor.protocol.api.ProtocolBridge;
import ink.reactor.protocol.api.packet.PacketOutbound;

import java.util.Collection;

public interface PlayerConnection {
    void addPacketToQueue(final PacketOutbound packet);
    void addPacketToQueue(final Collection<PacketOutbound> packets);

    void sendPacket(final PacketOutbound packet);
    void sendPackets(final Collection<PacketOutbound> packets);

    ProtocolBridge getBridge();
    ConnectionState getState();

    void setBridge(final ProtocolBridge bridge);
    void changeState(final ConnectionState state);

    void disconnect();
    void disconnect(final String reason);
}

File: networking/api/src/main/kotlin/ink/reactor/network/api/player/PlayerConnection.kt
package ink.reactor.network.api.player

import ink.reactor.network.api.packet.CachedPacket
import ink.reactor.network.api.packet.Packet
import io.netty.resolver.InetSocketAddressResolver

/**
 * Represents a network connection to a player.
 */
interface PlayerConnection {
    /**
     * Sends a packet to this specific player connection.
     *
     * This is a convenience method for simple packet sending to a single player.
     * For sending the same packet to multiple players, consider using
     * [ink.reactor.network.api.packet.PacketsSender] for better performance.
     *
     * @param packet The packet to send.
     */
    fun sendPacket(packet: Packet)

    /**
     * Sends multiple packets to this specific player connection.
     *
     * This is a convenience method for simple packet sending to a single player.
     * For sending the same packets to multiple players, consider using
     * [ink.reactor.network.api.packet.PacketsSender] for better performance.
     *
     * @param packets The packets to send.
     */
    fun sendPackets(vararg packets: Packet)

    fun sendCachedPacket(packet: CachedPacket)

    /**
     * Disconnects the player with the specified reason.
     *
     * @param reason The disconnection reason displayed to the player.
     */
    fun disconnect(reason: String)

    fun getIp(): String

    fun isOnline(): Boolean
}
